src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferAsSpanOverSubstring.cs

```diff
@@ -0,0 +1,298 @@
+            {
+                var method = invocation.TargetMethod;
+
+                if (candidate.Parameters.Length != method.Parameters.Length)
+                    return -1;
+
+                int replacementCount = 0;
+                foreach (var parameter in candidate.Parameters)
+                {
+                    if (isSubstringLookup[parameter.Ordinal] && SymbolEqualityComparer.Default.Equals(parameter.Type, symbols.ReadOnlySpanOfCharType))
+                    {
+                        ++replacementCount;
+                        continue;
+                    }
+
+                    var oldParameter = method.Parameters[parameter.Ordinal];
+                    if (!SymbolEqualityComparer.Default.Equals(parameter.Type, oldParameter.Type))
+                    {
+                        return -1;
+                    }
+                }
+
+                return replacementCount;
+            }
+        }
+
+        private static IEnumerable<IMethodSymbol> GetAllAccessibleOverloadsAtInvocationCallSite(IInvocationOperation invocation, CancellationToken cancellationToken)
+        {
+            var method = invocation.TargetMethod;
+            var model = invocation.SemanticModel;
+            int location = invocation.Syntax.SpanStart;
+            var instance = invocation.Instance;
+
+            IEnumerable<IMethodSymbol> allOverloads;
+            if (method.IsStatic)
+            {
+                allOverloads = model.LookupStaticMembers(location, method.ContainingType, method.Name).OfType<IMethodSymbol>();
+            }
+            //  Ensure protected members can only be invoked on instances that are known to be instances of the accessing class.
+            else if (instance is not null)
+            {
+                var enclosingType = GetEnclosingType(model, location, cancellationToken);
+                allOverloads = model.LookupSymbols(location, instance.Type, method.Name).OfType<IMethodSymbol>();
+                if (instance.Type.DerivesFrom(enclosingType, baseTypesOnly: true) || instance is IInstanceReferenceOperation)
+                {
+                    allOverloads = allOverloads.Union(model.LookupBaseMembers(location, method.Name).OfType<IMethodSymbol>());
+                }
+            }
+            //  This can happen when compiling invalid code.
+            else
+            {
+                allOverloads = Enumerable.Empty<IMethodSymbol>();
+            }
+
+            return allOverloads.Where(x => x.IsStatic == method.IsStatic && SymbolEqualityComparer.Default.Equals(x.ReturnType, method.ReturnType));
+
+            static INamedTypeSymbol GetEnclosingType(SemanticModel model, int location, CancellationToken cancellationToken)
+            {
+                ISymbol symbol = model.GetEnclosingSymbol(location, cancellationToken);
```

Comment:
Can this be null? If so, can `symbol.ContainingType` below throw a null reference exception?

Reply:
https://github.com/dotnet/roslyn-analyzers/blob/3f765296d9ea2fbcd88252a472e603b6c700e363/src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferAsSpanOverSubstring.cs#L190
I don't believe so. Even if the operation is a top-level statement, `GetEnclosingSymbol` should still return the `IMethodSymbol` representing the implicit entry-point. I'll add a test to ensure this is the case.